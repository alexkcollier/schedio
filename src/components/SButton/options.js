export {
  colors,
  allShadeOptions,
  colorNames,
} from '@/utils/component-options'

export const sizes = [
  'small',
  'regular',
  'large',
]

export const types = [
  'outlined',
  'text',
]
