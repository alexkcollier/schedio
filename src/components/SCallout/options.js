export const calloutTypes = ['', 'success', 'warning', 'error']
