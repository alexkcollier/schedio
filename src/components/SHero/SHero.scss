.hero {
  align-items: stretch;
  display: flex;
  flex-direction: column;
  justify-content: space-between;

  &--height {
    &-half {
      min-height: 50vh;
    }

    &-full {
      min-height: 100vh;
    }
  }

  &__body {
    align-items: center;
    display: flex;
    flex: 1 0 auto;
    padding: spacing('triple') spacing('loose');
  }

  &--has-no-padding {
    .hero__body {
      padding: spacing('none');
    }
  }

  &__foot,
  &__head {
    flex: 0 0 auto;
  }

  @each $color-name in $hero-colors {
    &--color-#{$color-name} {
      background-color: color($color-name);
      color: choose-contrast-color($color-name);
    }
  }
}
