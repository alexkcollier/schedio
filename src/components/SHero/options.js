export { colorNames as colors } from '@/utils/component-options';

export const heights = [
  'half',
  'two-thirds',
  'three-quarters',
  'almost-full',
  'full',
];
