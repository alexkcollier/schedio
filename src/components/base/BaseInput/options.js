export const allowed = ['email', 'text', 'password', 'url', 'search', ''].sort()
