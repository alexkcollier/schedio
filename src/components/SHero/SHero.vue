<template>
  <div
    :class="classList"
    class="hero"
  >
    <div
      v-if="$slots.head"
      class="hero__head"
    >
      <slot name="head" />
    </div>

    <div class="hero__body">
      <slot />
    </div>

    <div
      v-if="$slots.foot"
      class="hero__foot"
    >
      <slot name="foot" />
    </div>
  </div>
</template>

<script>
import * as options from './options'

const limitValues = propName => (v) => {
  const validOptions = options[`${propName}s`]

  return (
    !v ||
    validOptions.includes(v) ||
    console.error(`\`${propName}\` must be one of ${validOptions.join(', ')}`)
  )
}

export default {
  name: 'SHero',

  props: {
    height: {
      type: String,
      default: '',
      validator: limitValues('height'),
    },

    color: {
      type: String,
      default: '',
      validator: limitValues('color'),
    },

    hasNoPadding: {
      type: Boolean,
      default: false,
    },
  },

  computed: {
    classList () {
      return [
        this.height && `hero--height-${this.height}`,
        this.color && `hero--color-${this.color}`,
        this.hasNoPadding && 'hero--has-no-padding',
      ]
    },
  },

  created () {
    if (!this.$slots.default) {
      console.error(`\`${this.$options.name}\` requires default slot content.`)
    }
  },
}
</script>
