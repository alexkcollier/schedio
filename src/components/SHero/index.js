import SHero from './SHero.vue'

export { SHero }

export default {
  install: Vue => Vue.component(SHero.name, SHero)
}
