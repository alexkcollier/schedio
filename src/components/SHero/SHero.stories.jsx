import { SHero } from '@/components/SHero'
import { SHeading } from '@/components/SHeading'
import PropList from '@/docs/PropList'
import { boolean, select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'
import { colors, heights } from './options'

storiesOf('Components/Layout.SHero', module)
  .addParameters({ jest: 'SHero' })
  .add('SHero', () => {
    return {
      props: {
        color: {
          type: String,
          default: select('color', ['', ...colors], '', 'Optional Props')
        },
        height: {
          type: String,
          default: select('height', ['', ...heights], '', 'Optional Props')
        },
        hasNoPadding: {
          type: Boolean,
          default: boolean('has-no-padding', false, 'Optional Props')
        },
        slots: {
          default: () => ({
            head: text('head', 'Head slot', 'Slots'),
            default: text('default', 'Default slot', 'Slots'),
            foot: text('foot', 'Foot slot', 'Slots')
          })
        }
      },
      render(h) {
        const { slots, ...props } = this.$props
        return (
          <div>
            <SHeading level="1">SHero</SHeading>
            <p>
              Hero components are fullwidth banners. They are often the first visual element a user
              encounters on a page. They present an overview of a page's most important content.
              Padding on the default slot can be toggled with the <code>has-no-padding</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SHero {...{ props }}>
              {Object.entries(slots).map(slot => {
                return <template slot={slot[0]}>{slot[1]}</template>
              })}
            </SHero>

            {SHero.props && <PropList component={SHero} />}
          </div>
        )
      }
    }
  })
