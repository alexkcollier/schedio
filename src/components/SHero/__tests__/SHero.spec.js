import { shallowMount } from '@vue/test-utils'
import SHero from '@/components/SHero/SHero.vue'
import * as options from '@/components/SHero/options'
import { axe, toHaveNoViolations } from 'jest-axe'

expect.extend(toHaveNoViolations)

describe('SHero.vue', () => {
  let wrapper
  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => { })

  beforeEach(() => {
    wrapper = shallowMount(SHero, {
      slots: {
        head: 'Header Content',
        default: 'Body Content',
        foot: 'Foot Content',
      },
    })
  })

  afterEach(() => {
    jest.clearAllMocks()
  })

  it('renders correctly', async () => {
    expect(wrapper.html()).toMatchSnapshot()
    expect(await axe(wrapper.html())).toHaveNoViolations()
  })

  it('has a header', () => {
    expect(wrapper.vm.$slots.head).toBeTruthy()
  })

  it('has a body', () => {
    expect(wrapper.vm.$slots.default).toBeTruthy()
  })

  it('requires a default slot', () => {
    shallowMount(SHero)
    expect(errorSpy).toHaveBeenCalled()
  })

  it('hides empty slots', () => {
    const empty = shallowMount(SHero, { slots: { default: 'content' } })
    expect(empty.html()).toMatchSnapshot()
  })

  it('has a footer', () => {
    expect(wrapper.vm.$slots.foot).toBeTruthy()
  })

  options.colors.forEach((color) => {
    it(`can be ${color}`, async () => {
      wrapper.setProps({ color })
      expect(wrapper.html()).toMatchSnapshot()
      expect(await axe(wrapper.html())).toHaveNoViolations()
    })
  })

  options.heights.forEach((height) => {
    it(`can be ${height}`, () => {
      wrapper.setProps({ height })
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  it('can have no padding on the body', () => {
    wrapper.setProps({ hasNoPadding: true })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be any html tag', () => {
    wrapper.setProps({ tag: 'section' })
    expect(wrapper.contains('section')).toBe(true)
    expect(wrapper.html()).toMatchSnapshot()
  })
})
