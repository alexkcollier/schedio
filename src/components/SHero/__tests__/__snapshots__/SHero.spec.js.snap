// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHero.vue can be almost-full 1`] = `
<div class="hero hero--height-almost-full">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be any html tag 1`] = `
<section class="hero">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</section>
`;

exports[`SHero.vue can be full 1`] = `
<div class="hero hero--height-full">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be gold 1`] = `
<div class="hero hero--color-gold">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be green 1`] = `
<div class="hero hero--color-green">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be grey 1`] = `
<div class="hero hero--color-grey">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be half 1`] = `
<div class="hero hero--height-half">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be ice 1`] = `
<div class="hero hero--color-ice">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be magenta 1`] = `
<div class="hero hero--color-magenta">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be night 1`] = `
<div class="hero hero--color-night">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be red 1`] = `
<div class="hero hero--color-red">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be spartan_blue 1`] = `
<div class="hero hero--color-spartan_blue">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be three-quarters 1`] = `
<div class="hero hero--height-three-quarters">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be turquoise 1`] = `
<div class="hero hero--color-turquoise">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be two-thirds 1`] = `
<div class="hero hero--height-two-thirds">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can be white 1`] = `
<div class="hero hero--color-white">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue can have no padding on the body 1`] = `
<div class="hero hero--has-no-padding">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;

exports[`SHero.vue hides empty slots 1`] = `
<div class="hero">
  <!---->
  <div class="hero__body">content</div>
  <!---->
</div>
`;

exports[`SHero.vue renders correctly 1`] = `
<div class="hero">
  <div class="hero__head">Header Content</div>
  <div class="hero__body">Body Content</div>
  <div class="hero__foot">Foot Content</div>
</div>
`;
